app/src/promql/binOp.ts

```diff
@@ -0,0 +1,371 @@
+  [
+    binaryOperatorType.add,
+    binaryOperatorType.sub,
+    binaryOperatorType.mul,
+    binaryOperatorType.div,
+    binaryOperatorType.pow,
+    binaryOperatorType.mod,
+    binaryOperatorType.atan2,
+  ].includes(op);
+
+// Compute the time series labels for the result metric.
+export const resultMetric = (lhs: Metric, rhs: Metric, op: binaryOperatorType, matching: VectorMatching): Metric => {
+  const result: Metric = {};
+
+  // Start out with all labels from the LHS.
+  for (const name in lhs) {
+    result[name] = lhs[name];
+  }
+
+  // Drop metric name for operations that change the metric's meaning.
+  if (shouldDropMetricName(op)) {
+    delete result.__name__;
+  }
+
+  // Keep only match group labels for 1:1 matches.
+  if (matching.card === vectorMatchCardinality.oneToOne) {
+    if (matching.on) {
+      // Drop all labels that are not in the "on" clause.
+      for (const name in result) {
+        if (!matching.labels.includes(name)) {
+          delete result[name];
+        }
+      }
+    } else {
+      // Drop all labels that are in the "ignoring" clause.
+      for (const name of matching.labels) {
+        delete result[name];
+      }
+    }
+  }
+
+  // Include extra labels from the RHS that were mentioned in a group_x(...) modifier.
+  matching.include.forEach((name) => {
+    if (name in rhs) {
+      result[name] = rhs[name];
+    } else {
+      // If we are trying to include a label from the "one" side that is not actually set there,
+      // we need to make sure that we don't accidentally take its value from the "many" side
+      // if it exists there.
+      //
+      // Example to provoke this case:
+      //
+      // up == on(job, instance) group_left(__name__) node_exporter_build_info*1
+      delete result[name];
+    }
+  });
+
+  return result;
+};
+
+// Compute the match groups and results for each match group for a binary operator between two vectors.
+// In the error case, the match groups are still populated and returned, but the error field is set for
+// the respective group. Results are not populated for error cases, since especially in the case of a
+// many-to-many matching, the cross-product output can become prohibitively expensive.
+export const computeVectorVectorBinOp = (
+  op: binaryOperatorType,
+  matching: VectorMatching,
+  bool: boolean,
+  lhs: InstantSample[],
+  rhs: InstantSample[],
+  limits?: {
+    maxGroups?: number;
+    maxSeriesPerGroup?: number;
+  }
+): BinOpResult => {
+  // For the simplification of further calculations, we assume that the "one" side of a one-to-many match
+  // is always the right-hand side of the binop and swap otherwise to ensure this. We swap back in the end.
+  [lhs, rhs] = matching.card === vectorMatchCardinality.oneToMany ? [rhs, lhs] : [lhs, rhs];
+
+  const groups: BinOpMatchGroups = {};
+  const sigf = signatureFunc(matching.on, matching.labels);
+
+  const outOfLimitGroups = new Set<string>();
```

Comment:
you could just have an integer variable here to increase each time you would drop something.
It doesn't look like you are using this set except for its size

Reply:
ah yes you are right. Perhaps you could add a little comment to avoid the mistake later

Reply:
Done!